### PR TITLE
feat: support monorepo repo

### LINKED ISSUE #143: Support monorepo projects with pnpm
### Issues

- [X] I have checked [existing issues](https://github.com/nvimtools/none-ls.nvim/issues) and there are no existing ones with the same request.

### Feature description

In monorepo, the eslint is installed at the root `node_modules` by pnpm. 

In code https://github.com/nvimtools/none-ls.nvim/blob/8691504118b252d64fc5023a104aedd100ab754a/lua/null-ls/helpers/command_resolver.lua#L40

The `u.get_root()` will cause the searching only traverse up to the sub package's root if the nvim user are using some project manage plugin.

For example:

```
-- monorepo
  - node_modules
  - packages
     - foo-package
       - index.js
       - node_modules
```

If we are edit the index.js in foo-package, the `vim.uv.cwd()` will be `monorepo/packages/foo-package`, but the `eslint` are in `monorepo/node_modules/.bin` because of pnpm.

So `u.get_root()` returns `monorepo/packages/foo-package` and in https://github.com/nvimtools/none-ls.nvim/blob/8691504118b252d64fc5023a104aedd100ab754a/lua/null-ls/helpers/command_resolver.lua#L11 the search will be stopped immediately. 

### Help

Yes

### Implementation help

I suggest add an option to change the `end_path` when searching for command executable. for example, `config.get_search_root_stop_path`

### PR DESCRIPTION
**Pull request by towry**
resolves #143 

1. Just an idea.
2. The traverse logic may can be improved by first check the project root.

### CODE REVIEW

File: lua/null-ls/utils/init.lua [hunk 1]
```diff
@@ -211,6 +211,20 @@ M.handle_function_opt = function(opt, ...)
     return vim.deepcopy(opt)
 end
 
+--- gets git root
+---@return string
+M.get_git_root = function()
+    local result = vim.system({
```
Comment (mochaaP) [inline lua/null-ls/utils/init.lua:217]: prefer `vim.loop.spawn` here?

File: lua/null-ls/utils/init.lua [hunk 2]
```diff
@@ -211,6 +211,20 @@ M.handle_function_opt = function(opt, ...)
...
+        "git",
```
Comment (mochaaP) [inline lua/null-ls/utils/init.lua:218]: consider explicitly set search directory with `git -C ${M.get_root()}` (and likewise for other VCS) or set cwd on launch.

File: lua/null-ls/utils/init.lua [hunk 3]
```diff
@@ -211,6 +211,20 @@ M.handle_function_opt = function(opt, ...)
     return vim.deepcopy(opt)
 end
 
+--- gets git root
+---@return string
+M.get_git_root = function()
```
Comment (mochaaP) [inline lua/null-ls/utils/init.lua:216]: should be generic: `get_vcs_root`

**Review by mochaaP — CHANGES_REQUESTED** [review]

File: lua/null-ls/utils/init.lua (hunk 1 above)
Comment (mochaaP) [reply lua/null-ls/utils/init.lua:217]: also: traverse filesystem first for vcs directories (`.git`, `.svn`, `.bzr`, etc.) before spawning a new process.

Comment (towry) [reply lua/null-ls/utils/init.lua]: > prefer `vim.loop.spawn` here?

Seems spawn is async func, changed to `vim.fn.system`.

Comment (towry) [reply lua/null-ls/utils/init.lua]: traverse is added.

File: lua/null-ls/utils/init.lua (hunk 2 above)
Comment (towry) [reply lua/null-ls/utils/init.lua]: done and tested on local monorepo project. both local bin and monorepo root bin can be found.

File: lua/null-ls/utils/init.lua [hunk 4]
```diff
@@ -353,4 +353,23 @@ M.cosmiconfig = function(...)
...
+---@return string|nil
+M.get_vcs_root = function()
+    local uv = vim.uv or vim.loop
+    local cwd = uv.cwd()
+    ---@diagnostic disable-next-line: param-type-mismatch
+    local vcs_root = M.root_pattern(".git", ".hg", ".svn", ".bzr")(cwd)
+    if vcs_root then
+        return vcs_root
+    end
+    --- git worktree dir doesn't contains `.git` folder.
```
Comment (mochaaP) [inline lua/null-ls/utils/init.lua:366]: worktree checkouts have a `.git` file pointing to the bare repo. not sure if this is really needed

**Review by mochaaP — COMMENTED** [review]
aside from the one nitpick, it looks good to me. thanks!

File: lua/null-ls/utils/init.lua (hunk 4 above)
Comment (towry) [reply lua/null-ls/utils/init.lua:366]: Just double checked, you are right, I think this is not needed.

Comment (towry) [reply lua/null-ls/utils/init.lua]: Removed the system call.

**Review by mochaaP — APPROVED** [review]

**Pull request merged**